File: rs/moq/src/path.rs
use std::fmt::{self, Display};

/// A broadcast path that provides safe prefix matching operations.
///
/// This type wraps a String but provides path-aware operations that respect
/// delimiter boundaries, preventing issues like "foo" matching "foobar".
#[derive(Clone, Debug, PartialEq, Eq, Hash, Default)]
pub struct Path(String);

impl Path {
	/// Create a new Path from a string.
	pub fn new(s: impl Into<String>) -> Self {
		Self(s.into())
	}

	/// Check if this path has the given prefix, respecting path boundaries.
	///
	/// Unlike String::starts_with, this ensures that "foo" does not match "foobar".
	/// The prefix must either:
	/// - Be exactly equal to this path
	/// - Be followed by a '/' delimiter
	/// - Be empty (matches everything)
	///
	/// # Examples
	/// ```
	/// use moq_lite::Path;
	///
	/// let path = Path::new("foo/bar");
	/// assert!(path.has_prefix(&Path::new("foo")));
	/// assert!(path.has_prefix(&Path::new("foo/")));
	/// assert!(!path.has_prefix(&Path::new("fo")));
	///
	/// let path = Path::new("foobar");
	/// assert!(!path.has_prefix(&Path::new("foo")));
	/// ```
	pub fn has_prefix(&self, prefix: &Path) -> bool {
		if prefix.0.is_empty() {
			return true;
		}

		if !self.0.starts_with(&prefix.0) {
			return false;
		}

		// Check if the prefix is the exact match
		if self.0.len() == prefix.0.len() {
			return true;
		}

		// If the prefix ends with '/', it's already a proper boundary
		if prefix.0.ends_with('/') {
			return true;
		}

		// Otherwise, ensure the character after the prefix is a delimiter
		self.0.chars().nth(prefix.0.len()) == Some('/')
	}

	/// Strip the given prefix from this path, returning the suffix.
	///
	/// Returns None if the prefix doesn't match according to has_prefix rules.
	///
	/// # Examples
	/// ```
	/// use moq_lite::Path;
	///
	/// let path = Path::new("foo/bar/baz");
	/// let suffix = path.strip_prefix(&Path::new("foo")).unwrap();
	/// assert_eq!(suffix.as_str(), "/bar/baz");
	///
	/// let suffix = path.strip_prefix(&Path::new("foo/")).unwrap();
	/// assert_eq!(suffix.as_str(), "bar/baz");
	/// ```
	pub fn strip_prefix(&self, prefix: &Path) -> Option<Path> {
		if !self.has_prefix(prefix) {
			return None;
		}

		let suffix = &self.0[prefix.0.len()..];
		Some(Path(suffix.to_string()))
	}

	/// Get the path as a string slice.
	pub fn as_str(&self) -> &str {
		&self.0
	}

	/// Check if the path is empty.
	pub fn is_empty(&self) -> bool {
		self.0.is_empty()
	}

	/// Get the length of the path in bytes.
	pub fn len(&self) -> usize {
		self.0.len()
	}

	/// Join this path with another path component.
	///
	/// # Examples
	/// ```
	/// use moq_lite::Path;
	///
	/// let base = Path::new("foo");
	/// let joined = base.join("bar");
	/// assert_eq!(joined.as_str(), "foo/bar");
	///
	/// let base = Path::new("foo/");
	/// let joined = base.join("bar");
	/// assert_eq!(joined.as_str(), "foo/bar");
	/// ```
	pub fn join(&self, component: &str) -> Path {
		if self.0.is_empty() {
			Path::new(component)
		} else if self.0.ends_with('/') {
			Path::new(format!("{}{}", self.0, component))
		} else {
			Path::new(format!("{}/{}", self.0, component))
		}
	}
}

impl Display for Path {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		write!(f, "{}", self.0)
	}
}

impl AsRef<str> for Path {
	fn as_ref(&self) -> &str {
		&self.0
	}
}

impl From<String> for Path {
	fn from(s: String) -> Self {
		Self::new(s)
	}
}

impl From<&str> for Path {
	fn from(s: &str) -> Self {
		Self::new(s)
	}
}

impl From<&String> for Path {
	fn from(s: &String) -> Self {
		Self::new(s)
	}
}

impl From<&Path> for Path {
	fn from(p: &Path) -> Self {
		p.clone()
	}
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_has_prefix() {
		let path = Path::new("foo/bar/baz");

		// Valid prefixes
		assert!(path.has_prefix(&Path::new("")));
		assert!(path.has_prefix(&Path::new("foo")));
		assert!(path.has_prefix(&Path::new("foo/")));
		assert!(path.has_prefix(&Path::new("foo/bar")));
		assert!(path.has_prefix(&Path::new("foo/bar/")));
		assert!(path.has_prefix(&Path::new("foo/bar/baz")));

		// Invalid prefixes - should not match partial components
		assert!(!path.has_prefix(&Path::new("f")));
		assert!(!path.has_prefix(&Path::new("fo")));
		assert!(!path.has_prefix(&Path::new("foo/b")));
		assert!(!path.has_prefix(&Path::new("foo/ba")));
		assert!(!path.has_prefix(&Path::new("foo/bar/ba")));

		// Edge case: "foobar" should not match "foo"
		let path = Path::new("foobar");
		assert!(!path.has_prefix(&Path::new("foo")));
		assert!(path.has_prefix(&Path::new("foobar")));
	}

	#[test]
	fn test_strip_prefix() {
		let path = Path::new("foo/bar/baz");

		assert_eq!(path.strip_prefix(&Path::new("")).unwrap().as_str(), "foo/bar/baz");
		assert_eq!(path.strip_prefix(&Path::new("foo")).unwrap().as_str(), "/bar/baz");
		assert_eq!(path.strip_prefix(&Path::new("foo/")).unwrap().as_str(), "bar/baz");
		assert_eq!(path.strip_prefix(&Path::new("foo/bar")).unwrap().as_str(), "/baz");
		assert_eq!(path.strip_prefix(&Path::new("foo/bar/")).unwrap().as_str(), "baz");
		assert_eq!(path.strip_prefix(&Path::new("foo/bar/baz")).unwrap().as_str(), "");

		// Should fail for invalid prefixes
		assert!(path.strip_prefix(&Path::new("fo")).is_none());
		assert!(path.strip_prefix(&Path::new("bar")).is_none());
	}

	#[test]
	fn test_join() {
		assert_eq!(Path::new("foo").join("bar").as_str(), "foo/bar");
		assert_eq!(Path::new("foo/").join("bar").as_str(), "foo/bar");
		assert_eq!(Path::new("").join("bar").as_str(), "bar");
		assert_eq!(Path::new("foo/bar").join("baz").as_str(), "foo/bar/baz");
	}

	#[test]
	fn test_empty() {
		let empty = Path::new("");
		assert!(empty.is_empty());
		assert_eq!(empty.len(), 0);

		let non_empty = Path::new("foo");
		assert!(!non_empty.is_empty());
		assert_eq!(non_empty.len(), 3);
	}

	#[test]
	fn test_from_conversions() {
		let path1 = Path::from("foo/bar");
		let path2 = Path::from(String::from("foo/bar"));
		let s = String::from("foo/bar");
		let path3 = Path::from(&s);

		assert_eq!(path1.as_str(), "foo/bar");
		assert_eq!(path2.as_str(), "foo/bar");
		assert_eq!(path3.as_str(), "foo/bar");
	}
}

File: js/moq/src/path.ts
/**
 * A broadcast path that provides safe prefix matching operations.
 *
 * This class wraps a string but provides path-aware operations that respect
 * delimiter boundaries, preventing issues like "foo" matching "foobar".
 *
 * Paths are automatically trimmed of leading and trailing slashes on creation,
 * making all slashes implicit at boundaries.
 * All paths are RELATIVE; you cannot join with a leading slash to make an absolute path.
 *
 * @example
 * ```typescript
 * // Creation automatically trims slashes
 * const path1 = new Path("/foo/bar/");
 * const path2 = new Path("foo/bar");
 * console.log(path1.equals(path2)); // true
 *
 * // Safe prefix matching
 * const base = new Path("api/v1");
 * console.log(base.hasPrefix("api")); // true
 * console.log(base.hasPrefix(new Path("api/v1"))); // true
 *
 * const joined = base.join("users");
 * console.log(joined.toString()); // "api/v1/users"
 * ```
 */
export class Path {
	#path: string;

	/**
	 * Create a new Path from a string or another Path.
	 *
	 * Leading and trailing slashes are automatically trimmed.
	 */
	constructor(path: string | Path) {
		if (path instanceof Path) {
			// Trust the other Path instance to have already been normalized.
			this.#path = path.#path;
		} else {
			// Remove leading and trailing slashes, and collapse multiple slashes into one.
			this.#path = path.replace(/\/+/g, '/').replace(/^\/+/, '').replace(/\/+$/, '');
		}
	}

	/**
	 * Check if this path has the given prefix, respecting path boundaries.
	 *
	 * Unlike String.startsWith, this ensures that "foo" does not match "foobar".
	 * The prefix must either:
	 * - Be exactly equal to this path
	 * - Be followed by a '/' delimiter in the original path
	 * - Be empty (matches everything)
	 *
	 * @example
	 * ```typescript
	 * const path = new Path("foo/bar");
	 * console.log(path.hasPrefix("foo")); // true
	 * console.log(path.hasPrefix(new Path("foo"))); // true
	 * console.log(path.hasPrefix("foo/")); // true (trailing slash ignored)
	 * console.log(path.hasPrefix("fo")); // false
	 *
	 * const path2 = new Path("foobar");
	 * console.log(path2.hasPrefix("foo")); // false
	 * ```
	 */
	hasPrefix(prefix: string | Path): boolean {
		const prefixStr = new Path(prefix).#path;

		if (prefixStr === '') {
			return true;
		}

		if (!this.#path.startsWith(prefixStr)) {
			return false;
		}

		// Check if the prefix is the exact match
		if (this.#path.length === prefixStr.length) {
			return true;
		}

		// Otherwise, ensure the character after the prefix is a delimiter
		return this.#path[prefixStr.length] === '/';
	}

	/**
	 * Strip the given prefix from this path, returning the suffix.
	 *
	 * Returns null if the prefix doesn't match according to hasPrefix rules.
	 *
	 * @example
	 * ```typescript
	 * const path = new Path("foo/bar/baz");
	 * const suffix = path.stripPrefix("foo");
	 * console.log(suffix?.toString()); // "bar/baz"
	 *
	 * const suffix2 = path.stripPrefix(new Path("foo/"));
	 * console.log(suffix2?.toString()); // "bar/baz"
	 *
	 * const noMatch = path.stripPrefix("notfound");
	 * console.log(noMatch); // null
	 * ```
	 */
	stripPrefix(prefix: string | Path): Path | null {
		const prefixPath = new Path(prefix);
		if (!this.hasPrefix(prefixPath)) {
			return null;
		}

		// Handle empty prefix case
		if (prefixPath.#path === '') {
			return new Path(this);
		}

		// For non-empty prefix, skip the prefix and the following slash
		const suffix = Path.empty();
		suffix.#path = this.#path.slice(prefixPath.length + 1);
		return suffix;
	}

	/**
	 * Join this path with another path component.
	 *
	 * @example
	 * ```typescript
	 * const base = new Path("foo");
	 * const joined = base.join("bar");
	 * console.log(joined.toString()); // "foo/bar"
	 *
	 * const joined2 = base.join(new Path("bar"));
	 * console.log(joined2.toString()); // "foo/bar"
	 * ```
	 */
	join(other: string | Path): Path {
		const otherPath = new Path(other);

		if (this.#path === '') {
			return otherPath;
		} else if (otherPath.#path === '') {
			return this;
		} else {
			// Since paths are trimmed, we always need to add a slash
			// We avoid performing the regex; we know this is legit.
			const joined = Path.empty();
			joined.#path = `${this.#path}/${otherPath.#path}`;
			return joined;
		}
	}

	/**
	 * Check if this path is empty.
	 */
	isEmpty(): boolean {
		return this.#path === '';
	}

	/**
	 * Get the length of the path in characters.
	 */
	get length(): number {
		return this.#path.length;
	}

	/**
	 * Get the path as a string.
	 */
	toString(): string {
		return this.#path;
	}

	/**
	 * Check if this path equals another path.
	 */
	equals(other: string | Path): boolean {
		const otherPath = new Path(other);
		return this.#path === otherPath.#path;
	}

	/**
	 * Convert to JSON representation.
	 */
	toJSON(): string {
		return this.#path;
	}

	/**
	 * Static method to create an empty path.
	 */
	static empty(): Path {
		return new Path('');
	}
}
